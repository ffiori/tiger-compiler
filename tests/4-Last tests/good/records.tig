/* define valid recursive types */

let
  /* define a list */
  type intlist = {hd : int, tl : intlist}

  var l1 : intlist := intlist { hd = 1, tl = nil }
  var l2 : intlist := intlist { hd = 2, tl = l1 }
  var l3 : intlist := intlist { hd = 3, tl = l2 }
  var l4 : intlist := intlist { hd = 4, tl = l3 }

  function nth(n : int, list:intlist): int =
        (
		(while (n>0) do
		   (list:=list.tl;n:=n-1));
		list.hd
	)

  function printList(list:intlist) =
	if (list<>nil) then (printInt(list.hd);printList(list.tl))

  function plusOne(list:intlist) =
	if (list<>nil) then (list.hd:= list.hd+1;plusOne(list.tl))
in
	print("List elements, elementwise\n");
	printInt(nth(0,l4));
	printInt(nth(1,l4));
	printInt(nth(2,l4));
	printInt(nth(3,l4)); 
	print("List elements, whole list \n");
        printList(l4);
	print("List elements after mapping plus1\n");
	plusOne(l4);
	printList(l4);
        0
end


