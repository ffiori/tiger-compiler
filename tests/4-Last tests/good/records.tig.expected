List elements, elementwise
INT: 4
INT: 3
INT: 2
INT: 1
List elements, whole list 
INT: 4
INT: 3
INT: 2
INT: 1
List elements after mapping plus1
INT: 5
INT: 4
INT: 3
INT: 2
